### ISSUE TITLE
Frida refusing to attach on some MacOS Processes

### ISSUE THREAD

**Issue by mattshockl** [frida/frida#1829]
Hello,

I'm new to Frida and trying to follow some Google Project Zero tutorials of using this to intercept iMessages on MacOS Big Sur (https://github.com/googleprojectzero/iOS-messaging-tools/tree/master/iMessage).  I have SIP disabled, but whenever I try to use Frida to attach to imagent or apsd, I get the following error:

"Failed to attach: process with pid \<pid\> either refused to load frida-agent, or terminated during injection"

![screenshot](https://user-images.githubusercontent.com/5934002/131226628-3b314307-6e84-4ba5-85a5-32643051052e.png)

Am I just doing something completely wrong or is this a known issue?

**Comment by mattshockl** [comment]
Just realized it is injecting into the target process, but then immediately prints "Unable to fetch file descriptor from re.frida.piped.\<pid\>" to the target process's stdout...

**Comment by mattshockl** [comment]
This seems to be a sandboxing issue on MacOS/iOS with the mach-lookup entitlement.

On my Mac for apsd, /System/Library/Sandbox/Profiles/apsd.sb contains:
```
(allow mach-lookup
       (global-name-prefix "com.apple.aps.")
       ...
)
```

When creating the Mach ports for IPC, [Frida hardcodes the port name as re.frida.piped.\<pid\>](https://github.com/frida/frida-core/blob/master/src/darwin/frida-helper-backend-glue.m#L802):
```
  ctx = g_slice_new (FridaHelperContext);
  ctx->dispatch_queue = dispatch_queue_create ("re.frida.helper.queue", DISPATCH_QUEUE_SERIAL);

  ctx->piped_name = g_strdup_printf ("re.frida.piped.%u", getpid ());
  mach_port_allocate (mach_task_self (), MACH_PORT_RIGHT_RECEIVE, &ctx->piped_port);
```

The Frida agent then [attempts to open this mach port](https://github.com/frida/frida-core/blob/master/lib/pipe/pipe-darwin.c#L223):
``` 
  kr = bootstrap_look_up (bootstrap_port, service, &server);
  CHECK_MACH_RESULT (kr, ==, KERN_SUCCESS, "bootstrap_look_up");
```

Because `re.frida.piped` isn't in the sandbox profile for apsd, the Kernel prevents frida-agent from opening this mach port and returns the error code 1100 (BOOTSTRAP_NOT_PRIVILEGED, /usr/include/bootstrap.h).

To fix this issue temporarily, I changed the strings in frida-helper-backend-glue.m from `re.frida.helper.queue` and `re.frida.piped.<pid>` to `com.apple.aps.queue` and `com.apple.aps.piped.<pid>`, rebuilt Frida, and everything starting working perfectly.

I will attempt to think of a more elegant way to solve this problem in the coming days, but here is a down and dirty solution for anyone else struggling with this issue (@awelc, #1830).

**Comment by oleavr** [comment]
@mattshockl Thanks for digging into this!  One thing we could try is to fall back to using TCP over the loopback interface. The dynamically allocated port number could be included in the address string and used as a fallback. But hopefully there's a better solution right in front of us 🤔

**Comment by mattshockl** [comment]
Maybe you have more insight into the internals of these functions, but the [sandbox extension functions](https://github.com/frida/frida-core/blob/master/src/darwin/frida-helper-backend-glue.m#L584) were generating good tokens and not returning any errors when I was debugging yesterday.  And when the agent [consumed the token](https://github.com/frida/frida-core/blob/master/lib/pipe/pipe-darwin.c#L228), it returned a valid handle, but bootstrap_look_up still failed. I'm wondering if apple has some undocumented procedure for ignoring sandbox extensions for apple signed binaries or binaries running from launchd even with SIP disabled?

**Comment by oleavr** [comment]
@mattshockl Aha, interesting. Yeah that does sound like a plausible explanation – but unfortunately I haven't dug into this.

**Comment by awelc** [comment]
I tried the proposed solution but it still does not work work with the iOS app running on the simulator. @mattshockl - you obviously know a lot more about the system (I am really just a newbie...), so perhaps you have some additional ideas where to look for a solution.

After I modified frida-helper-backend-glue.m (I also confirmed that my /System/Library/Sandbox/Profiles/apsd.sb file contains the same "allow" directive), I am getting a subtly different error message when trying to attach to the process representing iOS app running on the simulator.

Previously this is what I was getting:
```
Frida: agent.vala:140: Unable to start agent: Unable to fetch file descriptor from re.frida.piped.23068 (bootstrap_look_up returned 'unknown error code')
```

And now this is what I am getting (please not the error code is different and now talks about "unknown service name"):
```
Frida: agent.vala:140: Unable to start agent: Unable to fetch file descriptor from com.apple.aps.piped.81071 (bootstrap_look_up returned 'Unknown service name')
```

**Comment by mattshockl** [comment]
Hello, that is occurring because Frida is registering "com.apple.aps.piped" as a bootstrap service in the user's namespace and apsd is running in the system namespace under launchd and cannot see the user namespace services.  If you ran apsd from the terminal directly and then tried to use Frida on it in a different window, this fix would work.  To make it work with launchd services, I recommend using the temporary solution in https://github.com/frida/frida-core/pull/392 until a better solution is figured out.

**Comment by oleavr** [comment]
This issue is now fixed in Frida 15.1.0.

Note however that imagent's sandbox profile is a bit too strict on Monterey, so attempting to attach to it will crash it because we currently rely on the `pipe()` syscall. The solution to this is probably patching GLib to use `kqueue()` instead of `poll()`, so we can wake up the `kwait()` using a Mach port. Anyway, that's a challenge for another time.
